### PR TITLE
498 Add list to dict utility to parse variables

### PR DESCRIPTION
**Pull request by Nic-Ma**
part of #498 

### Description
This PR implemented `list_to_dict` utility to parse variables from string config file.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/utils/misc.py [hunk 1]
```diff
@@ -217,3 +218,39 @@ def set_determinism(
...
+def list_to_dict(items):
+    """
+    To convert a list of "key=value" pairs into a dictionary.
+    For examples: items: `["a=1", "b=2", "c=3"]`, return: {"a": "1", "b": "2", "c": "3"}.
+    If no "=" in the pair, use None as the value, for example: ["a"], return: {"a": None}.
+
+    """
+
+    def _parse_var(s):
+        items = s.split("=")
```
Comment (ericspod) [inline monai/utils/misc.py:232]: Here you can say `s.split("=",1)` to split by = character only once, this would return 1 or 2 values exactly.

File: monai/utils/misc.py [hunk 2]
```diff
@@ -217,3 +218,39 @@ def set_determinism(
...
+            # rejoin the rest, for example: MODEL="checkpoint_best_metric=0.94.pth"
+            value = "=".join(items[1:])
+        return key, value
+
+    d = dict()
+    if items:
+        for item in items:
+            key, value = _parse_var(item)
+
+            try:
```
Comment (ericspod) [inline monai/utils/misc.py:246]: I would suggest `ast.literal_eval` would be a more robust solution.

File: monai/utils/misc.py [hunk 3]
```diff
@@ -217,3 +218,39 @@ def set_determinism(
         torch.backends.cudnn.benchmark = False
     else:
         torch.backends.cudnn.deterministic = False
+
+
+def list_to_dict(items):
```
Comment (ericspod) [inline monai/utils/misc.py:223]: The `configparser` library might make sense for this but if the config is really simple it's just as well to parse ourselves.

File: monai/utils/misc.py [hunk 4]
```diff
@@ -217,3 +218,39 @@ def set_determinism(
...
+    If no "=" in the pair, use None as the value, for example: ["a"], return: {"a": None}.
+
+    """
+
+    def _parse_var(s):
+        items = s.split("=")
+        # we remove blanks around keys, as is logical
+        key = items[0].strip()
+        value = None
+        if len(items) > 1:
```
Comment (wyli) [inline monai/utils/misc.py:236]: this could be implemented with `split("=", maxsplit=1)`? https://docs.python.org/3/library/stdtypes.html#str.split

should we detection the duplicated keys such as: `["a=1", "b=2", "a=3"]`?

Comment (Nic-Ma) [reply monai/utils/misc.py:236]: updated in the latest commit.
Thanks.

File: monai/utils/misc.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/utils/misc.py:232]: updated in the latest commit.
Thanks.

File: monai/utils/misc.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/utils/misc.py:223]: Hi @ericspod , thanks for your suggestion, we only use this utility to parse several key-value pairs in the config file.
If we have complicated objects to parse, we can try to use some 3rd party lib I think.
Thanks.

File: monai/utils/misc.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/utils/misc.py:246]: Hi @ericspod ,

I tried it and got some errors:
```
>>> print(literal_eval("false"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.6/ast.py", line 85, in literal_eval
    return _convert(node_or_string)
  File "/opt/conda/lib/python3.6/ast.py", line 84, in _convert
    raise ValueError('malformed node or string: ' + repr(node))
ValueError: malformed node or string: <_ast.Name object at 0x7f3820bed7b8>
>>> print(literal_eval("haha"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.6/ast.py", line 85, in literal_eval
    return _convert(node_or_string)
  File "/opt/conda/lib/python3.6/ast.py", line 84, in _convert
    raise ValueError('malformed node or string: ' + repr(node))
ValueError: malformed node or string: <_ast.Name object at 0x7f3820c0ad68>
```
Should we use `eval`?

Thanks.

Comment (ericspod) [reply monai/utils/misc.py:246]: `literal_eval` will parse only literal values in a safe way, so something like "x=3" won't parse. You can use `literal_eval` and resort to keeping the original string if this exception is raised rather than having the nested try/except blocks.

Comment (Nic-Ma) [reply monai/utils/misc.py:246]: But we also need to parse bool type as shown in the above test example.
So I still use the previous try-catch for bool and string and use `literal_eval ` for int and float?

Thanks.

Comment (Nic-Ma) [reply monai/utils/misc.py:246]: updated in the latest commit.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli and @ericspod ,

Thanks for your review, I updated the PR according to your comments.
Do you think we can merge it now?

Thanks.

**Comment by wyli** [comment]
> Hi @wyli and @ericspod ,
> 
> Thanks for your review, I updated the PR according to your comments.
> Do you think we can merge it now?
> 
> Thanks.

thanks, would be helpful to include more tests for the corner cases?
```
['a=1', 'a=2']
['a = 01 ']
['a=None']
['a=']
['a: 0']
["a='test'"]
["'a'=0"]
```
I'm not 100% sure about the intended use cases, so please ignore these if I'm wrong

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Add more tests for core cases.
Thanks.
